onnxruntime/core/providers/cpu/ml/ml_common.h

```diff
@@ -342,5 +356,72 @@ void write_scores(std::vector<T>& scores, POST_EVAL_TRANSFORM post_transform, in
   memcpy(out_p, scores.data(), len);
 }
 
+// TODO: Starting with just the pieces needed for LinearRegressor from write_scores (see above).
+//       Will see what can be sensibly added to a batched in-place update of the scores for LinearClassifier, the SVM*
+//       and TreeEnsemble* ops when updating those.
+//       Attempted to parallelize the calculations if the number of scores to process was large, but no clear benefit
+//       was seen from testing with the arbitrary values of 1000 scores per threads.
+template <typename T>
+void batched_update_scores_inplace(gsl::span<T> scores, int64_t num_batches64, int64_t batch_size,
+                                   POST_EVAL_TRANSFORM post_transform, int add_second_class,
+                                   concurrency::ThreadPool* threadpool) {
+  SafeInt<int32_t> num_batches(num_batches64);
+  SafeInt<int32_t> num_scores = num_batches * batch_size;
+
+  ORT_ENFORCE(scores.size() == static_cast<size_t>(num_scores));
+
+  ORT_UNUSED_PARAMETER(add_second_class);  // pending updates to use batched_update_scores_inplace for more ops
+  ORT_UNUSED_PARAMETER(threadpool);        // TBD whether we need to parallelize code here
+
+  // convert from span to pointer for efficiency. we've checked scores.size() matches num_scores so don't need the
+  // extra checking/overhead from using operator[] for each access
+  T* s = &*scores.begin();
+  const T* s_end = &*scores.cend();
+
+  if (batch_size > 1) {
```

Comment:
You can remove else if part by changing the if condition to:
```
if (batch_size > 1 || post_transform == POST_EVAL_TRANSFORM::PROBIT)
```
I assume batch size >= 1.

Reply:
I have the changes to convert LinearClassifier to use GEMM as well, and those add all the handling for 'add_second_class' in the 'else' (see line 332 - 345 above), so any removal here would be temporary. 

---
In reply to: [389724122](https://github.com/microsoft/onnxruntime/pull/3154#discussion_r389724122) [](ancestors = 389724122)